app/models/user.rb

```diff
@@ -146,6 +147,19 @@ def to_s
     name
   end
 
+  def validate_openai_api_key
+    return unless openai_api_key_changed?
+
+    client = OpenAI::Client.new(access_token: openai_api_key)
+
+    begin
+      response = client.models.list
+      errors.add(:base, :invalid_api_key) unless response['data']
+    rescue Faraday::UnauthorizedError, OpenAI::Error
+      errors.add(:base, :invalid_api_key)
+    end
```

Comment:
To reduce code duplication, I would prefer to have this code exclusively put in the `TaskService::GptGenerateTests`.

Can't we do something like 

```rb
TaskService::GptGenerateTests.validate_api_key! openai_api_key
```

here?

Reply:
What do you mean? where are you proposing to do this, in the `user.rb` ? and how can this be done? :)

Reply:
I tried to make it work but I couldn't. Also after some examination, I noticed they are not entirely duplicated. 
In the `user.rb` file I also check if the key has been changed or not to avoid checking it everytime a user updates their profile without changing their key. Also the way the error is raised in the `user.rb' file is different than the one used at the `gpt_generate_tests.rb` file. Simply calling the `validate_api_key` in the `gpt_generate_tests.rb` from the `user.rb` file will cause an error that is not handled in that case.

While the code does a similar function, it is not necessarily the same. I think its better to have them separated but my ruby knowledge is also limited with its capabilities. So maybe you are aware of a way to do it without causing the issues I am facing :)

Reply:
Thanks for trying, I appreciate that. The first part of the method (checking whether the key was changed), not marked in my comment, needs to stay in the `User` model.

I just pushed my proposal as two commits. Please have a look and test them.

Reply:
@MrSerth Thank you for you making the changes yourself. I appreciate you taking the time to do that. The generate test specs looks cleaner.

I have tested the API key change and I now receive an error when changing the key to an invalid one. `the server responded with status 503` caused by this line:
```
    def self.validate!(client)
      response = client.models.list
  ```
  
  Not sure what happened?

Reply:
After some investigation, it seems that there is currently an outage in ChatGPT and API assistance services. So the error is not from our side. (But of course, we now need to handle this case as well so the user sees a better error message). 🥹👀

Reply:
@MrSerth I just pushed a commit with a solution for this issue :) 
Please check it out and if it's satisfactory, then feel free to resolve this comment.

Reply:
p.s. I managed to test it quickly before openAI fixed the error and it worked.

Reply:
Thanks! I refactored the error handling a little further, to make it easier extendable and maintainable. Please have another look and feel free to resolve the thread 😉

Reply:
Thanks! this way it will be easier to add future error handling.  Thanks for your feedback and help :)